Update clang-tidy documentation with exceptions note (#258)
ros-planning/moveit.ros.org: documentation/contributing/code/index.markdown

catkin config --cmake-args -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_CXX_CLANG_TIDY=clang-tidy
catkin build
```

### Exceptions to clang-tidy

> In some cases, the warnings generated by clang-tidy may be false-positive, e.g. `performance-unnecessary-copy-initialization` may falsely convert a `non-const` variable to `const`, or cannot be silenced for some reason.

Reviewer: To be honest, I cannot confirm your observations. All changes suggested by clang-tidy look very good.
I reverted all of your exceptions without any issues. See https://github.com/ros-planning/moveit/pull/1376.

That's why I suggest dropping this paragraph.

Author: My first thought was that it may be the difference between clang-tidy 3.9 and 6.0. After I removed 3.9 on my disc, and run 6.0 from `clang-tidy clang-tools`, clang-tidy still converts the non-const `robot_state` to const ref. Then I further cleaned up my disc by running `apt remove clang* && apt autoremove` and install clang-tidy 6.0 again. This warning never came up and the rest warnings match the travis_ci result. Perhaps the previous version of clang related pkgs `libllvm` on my disc brings the difference.

Maybe this info and clang-tidy version should be clarified in the doc. Seems clang-tidy 6.0 do actually have more `performance-*` checks and more warnings with the same check than clang-tidy 3.9.

> In some cases, the warnings generated by clang-tidy may be false-positive, e.g. `performance-unnecessary-copy-initialization` may falsely convert `non-const` variable to `const` variable, or cannot be silenced for some reason.

Reviewer: ```suggestion
In some cases, the warnings generated by clang-tidy may be false-positive, e.g. `performance-unnecessary-copy-initialization` may falsely convert a `non-const` variable to `const`, or cannot be silenced for some reason.
```

> It is recommended to suppress undesired clang-tidy checks by using `NOLINT` or `NOLINTNEXTLINE` comments followed by the list of check names in parentheses:

Reviewer: ```suggestion
It is possible to suppress false-positive clang-tidy checks by using `NOLINT` or `NOLINTNEXTLINE` comments followed by the list of check names in parentheses:
```

Author: Here I use `undesired` because some cases may not be false-positive but not desired for some reason. By using `recommended`, I intends to recommend specifying the check names instead of `NOLINT` alone.

> It is recommanded to suppress the undesired clang-tidy checks by using `NOLINT` and `NOLINTNEXTLINE` comments followed by the list of check names in a parentheses:

Reviewer: ```suggestion
It is recommended to suppress undesired clang-tidy checks by using `NOLINT` or `NOLINTNEXTLINE` comments followed by the list of check names in parentheses:
```

```c++
const IKCallbackFn solution_callback = 0; // NOLINT(modernize-use-nullptr)

// NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
robot_state::RobotState robot_state(default_state);
```

> Note that `modernize-loop-convert` check may convert `for (...; ...; ...)` loops to `for (auto & ... : ...)`. However, `auto` is not an expression highly readable, please consider specify the variable type like:

Reviewer: ```suggestion
Note that `modernize-loop-convert` check may convert `for (...; ...; ...)` loops to `for (auto & ... : ...)`. However, `auto` is not an expression highly readable, please consider explicitly specifying the variable type like:
```
